File: truss-compiler/src/main/java/truss/compiler/symbols/Symbol.java
package truss.compiler.symbols;

import org.apache.commons.lang.Validate;
import truss.compiler.Span;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public abstract class Symbol {
    private final List<Span> spans = new ArrayList<>();
    private final List<Span> unmodifiableSpans = Collections.unmodifiableList(spans);

    protected void addSpan(Span span) {
        Validate.notNull(span, "span");

        spans.add(span);
    }

    public List<Span> getSpans() {
        return unmodifiableSpans;
    }

    public String getName() {
        return null;
    }

    public String getMetadataName() {
        return getName();
    }

    public abstract SymbolKind getKind();

    public abstract void accept(SymbolVisitor visitor);
}

File: Truss.Compiler/Symbols/Symbol.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;

namespace Truss.Compiler.Symbols {
    public abstract class Symbol {
        private readonly List<Span> _spans = new List<Span>();

        protected Symbol() {
            Spans = new ReadOnlyCollection<Span>(_spans);
        }

        protected void AddSpan(Span span) {
            if (span == null) {
                throw new ArgumentNullException("span");
            }

            _spans.Add(span);
        }

        public IList<Span> Spans { get; private set; }

        public virtual string Name {
            get { return null; }
        }

        public virtual string MetadataName {
            get { return Name; }
        }

        public abstract SymbolKind Kind { get; }

        public abstract void Accept(ISymbolVisitor visitor);
    }
}
